substrate/primitives/genesis-builder/src/lib.rs

```diff
@@ -19,39 +19,107 @@
+//! on some input arbitrary bytes array. This documentation uses term `RuntimeGenesisConfig`, which
+//! for non-FRAME runtimes may be understood as the runtime-side entity representing initial runtime
+//! configuration. The representation of the preset is an arbitrary `Vec<u8>` and does not
+//! necessarily have to represent a JSON blob.
+//!
+//! The runtime may provide a number of partial predefined `RuntimeGenesisConfig` configurations in
+//! the form of patches which shall be applied on top of the default `RuntimeGenesisConfig`. The
+//! patch is a JSON blob, which essentially comprises the list of key-value pairs that are to be
+//! customized in the default runtime genesis config. These predefined configurations are referred
+//! to as presets.
+//!
+//! This allows the runtime to provide a number of predefined configs (e.g. for different
+//! testnets or development) without neccessity to leak the runtime types outside the itself (e.g.
+//! node or chain-spec related tools).
+//!
+//! This Runtime API allows to interact with `RuntimeGenesisConfig`, in particular:
+//! - provide the list of available preset names,
+//! - provide a number of named presets of `RuntimeGenesisConfig`,
+//! - provide a JSON represention of the default `RuntimeGenesisConfig` (by simply serializing the
+//!   default `RuntimeGenesisConfig` struct into JSON format),
+//! - deserialize the full `RuntimeGenesisConfig` from given JSON blob and put the resulting
+//!   `RuntimeGenesisConfig` structure into the state storage creating the initial runtime's state.
+//!   Allows to build customized genesis. This operation internally calls `GenesisBuild::build`
+//!   function for all runtime pallets.
+//!
+//! Providing externalities with an empty storage and putting `RuntimeGenesisConfig` into storage
+//! (by calling `build_state`) allows to construct the raw storage of `RuntimeGenesisConfig`
+//! which is the foundation for genesis block.
 
 extern crate alloc;
+use alloc::vec::Vec;
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
 
 /// The result type alias, used in build methods. `Err` contains formatted error message.
 pub type Result = core::result::Result<(), sp_runtime::RuntimeString>;
 
+/// Wrapper type representing preset ID.
+///
+/// Preset ID length is limited to 64 bytes. For user convenience, it is recommended that the ID be
+/// `utf8` string, as some tooling may not support arbitrary bytes array.
+#[derive(Decode, Encode, Debug, TypeInfo, PartialEq)]
+pub struct PresetId(sp_runtime::BoundedVec<u8, sp_runtime::traits::ConstU32<64>>);
+
+impl From<&str> for PresetId {
+	fn from(v: &str) -> Self {
+		Self(
+			v.as_bytes()
+				.to_vec()
+				.try_into()
+				.expect("Preset name length shall be less than 64. qed."),
+		)
+	}
+}
+
+impl<'a> TryFrom<&'a PresetId> for &'a str {
+	type Error = core::str::Utf8Error;
+	fn try_from(v: &'a PresetId) -> core::result::Result<&'a str, Self::Error> {
+		core::str::from_utf8(&v.0)
+	}
+}
```

Comment:
Why isn't this just `RuntimeString`?

Reply:
Could be easily done with this extension: #3942 (initially I didn't want to add this).

Reply:
done in: c374a37
Maybe we don't need extra type at all....